docs/guides/0X_v8_migration_guide.md

```diff
@@ -0,0 +1,293 @@
+
+    return next(httpContext);
+}
+```
+
+\_NOTE: Since CQRS related middlewares are plain ASP.NET middleware, it's possible to register them outside of `MapRemoteCQRS` method. This will throw a runtime exception since `GetCQRSEndpoint()` and `GetCQRSRequestPayload()` require CQRS metadata to be present in `HttpContext` (see the chart below)
+
+#### CQRS Request handling
+
+```mermaid
+sequenceDiagram
+    participant aspnet as ASP.NET middlewares
+    participant start as CQRSMiddleware
+    participant middle as CQRS specific middlewares
+    participant final as CQRSPipelineFinalizer
+
+    Note over aspnet: Common ASP.NET middlewares, e.g. <br/> UseAuthentication, UseCors
+
+    aspnet ->> start: next()
+
+    Note over start: Deserialize request according <br/> to CQRSEndpointMetadata
+    Note over start: Set CQRSRequestPayload
+
+    start ->> middle: next()
+    Note over middle: Custom middlewares, e.g. validation, security
+
+    middle ->> final: next()
+
+    Note over final: Execute handler
+    Note over final: Set result in CQRSRequestPayload
+
+    final ->> middle: 
+    Note over middle: Custom middlewares, e.g. events publication
+
+    middle ->> start: 
+
+    Note over start: Set response headers
+    Note over start: Serialize result
+
+    start ->> aspnet: 
+
+```
+
+### Fluent validation changes
+
+The `ContextualValidator.RuleForAsync()` was removed due to problems with upgrading to newer `FluentValidation` versions and general impedance with _fluent_ approach. Instead clients should rewrite the validator to use imperative approach with `AbstractValidator.CustomAsync()`.
+
+There is a `IRuleBuilderOptions.AddValidationError()` extension which allows setting CQRS error code to the validation error.
+
+Example old approach:
+
+```csharp
+public class CreateDishCV : ContextualValidator<CreateDish>
+{
+    public CreateDishCV()
+    {
+        this.RuleForAsync(
+            c => c.DishId,
+            (ctx, id) => ctx.GetService<IRepository<Dish>>().FindAsync(id))
+            .Null().WithCode(CreateDish.ErrorCodes.DishAlreadyExists);
+    }
+}
+```
+
+Example new approach:
+
+```csharp
+public class CreateDishCV : ContextualValidator<CreateDish>
+{
+    public CreateDishCV()
+    {
+        RuleFor(c => c.DishId)
+            .CustomAsync((Guid dishId, ValidationContext<CreateDish> ctx, CancellationToken ct) =>
+            {
+                var dish = await ctx.GetService<IRepository<Dish>>().FindAsync(id, ct);
+                if(dish is not null)
+                {
+                  ctx.AddValidationError("Dish already exists", CreateDish.ErrorCodes.DishAlreadyExists);
+                }
+            });
+    }
+}
+```
+
+Removal of `ContextualValidator` class allowed to merge `LeanCode.Validation.Fluent` and `LeanCode.Validation.Fluent.Scoped` projects - lifetime of validators is now a parameter of `IServiceCollection.AddFluentValidation()` method (scoped by default)
```

Comment:
> Removal of `ContextualValidator`

But the "new approach" above still uses it in ` public class CreateDishCV : ContextualValidator<CreateDish>`? :thinking: